doc/codeExecution.md

```diff
@@ -0,0 +1,77 @@
+
+### `native` execution mode
+
+The `native` execution mode is currently used when PrairieLearn is running in production. It uses Docker to provide a degree of isolation from both PrairieLearn and other courses.
+
+Instead of using a pool of Python trampolines as described above, it actually maintains a pool of Docker containers, each of which runs a simple Node script, which in turn runs a Python trampoline. The Node script listens for requests from PrairieLearn and essentially just forwards them to the Python process. You may ask, "Why not just run the trampoline as the primary process in the container?" Well, starting up a Docker container is significantly more expensive than starting up a Python interpreter. Given that we ocasionally want to completely restart the Python worker, such as when it encounters an error, having an additional level of indirection allows us to gracefully restart the Python process inside the Docker container without having to restart the entire Docker container.
+
+This mode also allows us to isolate one course from another so that course A cannot see content from course B, and vice versa. To achieve this, we take advantage of bind mounts. When creating a container, PrairieLearn also creates a special directory on the host, and then mounts that directory to `/course` in the container. To execute content for course A, PrairieLearn first bind mounts that course's directory to the container's host directory. This is transitive to the container, which will now see that course's content at `/course`. In the future, when a different course's code needs to be executed in that container, PrairieLearn will simply update the bind mount to point to the other course.
+
+### `container` execution mode
+
+The `container` execution mode is similar to the `native` mode in that code is executed in a container, but is used where PrairieLearn is *also* running in a container. This is primarily useful for when you're running PrairieLearn locally but want to test running course code in isolated containers.
+
+In this case, we can't use the bind mount trick above, as PrairieLearn can't create mounts on the host from inside its container. Instead, we'll need a couple of things:
+
+* In `config.json`, the user must specify the host path for any courses they're mounting into PrairieLearn in `courseDirsHost`.
+* The user must mount the Docker socket into PrairieLearn's container, similarly to what they must do when running external grading locally.
+* The user must create a "scratch" directory on the host, specify it with the `HOSTFILES_DIR` environment variable, and mount the directory to `/hostfiles` in the PrairieLearn container. PriaireLearn will use this directory to communicate its internal files, including element implementations and the Python trampoline, to the host so that they can be mounted into executor containers.
```

Comment:
```suggestion
* The user must create a "scratch" directory on the host, specify it with the `HOSTFILES_DIR` environment variable, and mount the directory to `/hostfiles` in the PrairieLearn container. PrairieLearn will use this directory to communicate its internal files, including element implementations and the Python trampoline, to the host so that they can be mounted into executor containers.
```